bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/discovery/SunspecDiscoveryProcess.java

```diff
@@ -0,0 +1,397 @@
+import org.openhab.io.transport.modbus.ModbusReadFunctionCode;
+import org.openhab.io.transport.modbus.ModbusReadRequestBlueprint;
+import org.openhab.io.transport.modbus.ModbusRegisterArray;
+import org.openhab.io.transport.modbus.ModbusSlaveErrorResponseException;
+import org.openhab.io.transport.modbus.PollTask;
+import org.openhab.io.transport.modbus.endpoint.ModbusSlaveEndpoint;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is used by the SunspecDiscoveryParticipant to detect
+ * the model blocks defined by the given device.
+ * It scans trough the defined model items and notifies the
+ * discovery service about the discovered devices
+ *
+ * @author Nagy Attila Gabor - Initial contribution
+ */
+@NonNullByDefault
+public class SunspecDiscoveryProcess {
+
+    /**
+     * Logger instance
+     */
+    private final Logger logger = LoggerFactory.getLogger(SunspecDiscoveryProcess.class);
+
+    /**
+     * The handler instance for this device
+     */
+    private final ModbusEndpointThingHandler handler;
+
+    /**
+     * The endpoint where we can reach the device
+     */
+    private final ModbusSlaveEndpoint endpoint;
+
+    /**
+     * Listener for the discovered devices. We get this
+     * from the main discovery service, and it is used to
+     * submit any discovered Sunspec devices
+     */
+    private final ModbusDiscoveryListener listener;
+
+    /**
+     * The endpoint's slave id
+     */
+    private int slaveId;
+
+    /**
+     * Number of maximum retries
+     */
+    private static final int maxTries = 3;
+
+    /**
+     * List of start addresses to try
+     */
+    private List<Integer> possibleAddresses;
+
+    /**
+     * This is the base address where the next block should be searched for
+     */
+    private int baseAddress = 40000;
+
+    /**
+     * Count of valid Sunspec blocks found
+     */
+    private int blocksFound = 0;
+
+    /**
+     * Parser for commonblock
+     */
+    private final CommonModelParser commonBlockParser;
+
+    /**
+     * The last common block found. This is used
+     * to get the details of any found devices
+     */
+    private @Nullable CommonModelBlock lastCommonBlock = null;
+
+    /**
+     * New instances of this class should get a reference to the handler
+     *
+     * @throws EndpointNotInitializedException
+     */
+    public SunspecDiscoveryProcess(ModbusEndpointThingHandler handler, ModbusDiscoveryListener listener)
+            throws EndpointNotInitializedException {
+        this.handler = handler;
+
+        ModbusSlaveEndpoint endpoint = this.handler.asSlaveEndpoint();
+        if (endpoint != null) {
+            this.endpoint = endpoint;
+        } else {
+            throw new EndpointNotInitializedException();
+        }
+        slaveId = handler.getSlaveId();
+        this.listener = listener;
+        commonBlockParser = new CommonModelParser();
+        possibleAddresses = new CopyOnWriteArrayList<>();
```

Comment:
If you want a thread-safe queue you can use this:
```suggestion
        possibleAddresses = new ConcurrentLinkedQueue<>();
```
But I'm not entirely sure that you need a thread-safe queue implementation here anyway...

Reply:
Thanks I changed this.
The main idea here is to create a new instance of SunspecDiscoveryProcess for each discovery request, so it can keep track of the discovery messages. So concurrent access is highly unlikely but it's better to be safe than sorry.